### PR TITLE
Added the framework for the OCS rules and alerts

### PR DESCRIPTION
**Pull request by anmolsachan**
This PR adds a framework for generating OCS rules and alerts. 

This is a follow-up PR on #733, #737 and #739. This PR should clear up how OCS rules yamls are created and how they will be maintained in future.

Doc to install dependencies and use : metrics/mixin/README.md

Signed-off-by: Anmol Sachan <anmol13694@gmail.com>

### CODE REVIEW

**Comment by anmolsachan** [comment]
@umangachapagain @obnoxxx @jarrpa Please review.

**Comment by anmolsachan** [comment]
> I have literally no idea how any of this works. Please include documentation (say under `metrics/docs`) to explain what all this means and how developers are supposed to use this. Links to external tutorials are also valid but probably not sufficient.

@jarrpa I understand and thanks for pointing that out. Since it is related to only "MIXIN' I have added the readme at metrics/mixin/README.md

**Comment by jarrpa** [comment]
> > I have literally no idea how any of this works. Please include documentation (say under `metrics/docs`) to explain what all this means and how developers are supposed to use this. Links to external tutorials are also valid but probably not sufficient.
> 
> @jarrpa I understand and thanks for pointing that out. Since it is related to only "MIXIN' I have added the readme at metrics/mixin/README.md

That still doesn't tell me what any of this means or what it does. Also, it doesn't mention how to actually make changes, just hwo to generate artifacts after the actual work has been done.

**Comment by anmolsachan** [comment]
> > > I have literally no idea how any of this works. Please include documentation (say under `metrics/docs`) to explain what all this means and how developers are supposed to use this. Links to external tutorials are also valid but probably not sufficient.
> > 
> > 
> > @jarrpa I understand and thanks for pointing that out. Since it is related to only "MIXIN' I have added the readme at metrics/mixin/README.md
> 
> That still doesn't tell me what any of this means or what it does. Also, it doesn't mention how to actually make changes, just hwo to generate artifacts after the actual work has been done.

@jarrpa I have added more info in the README file. Should make things more clear.

File: metrics/mixin/README.md
Comment (jarrpa) [inline metrics/mixin/README.md:30]: This is not sufficient to even explain your example. For instance, I have no idea what `+::` does, as it is mentioned nowhere in the Jsonnet website. You can't assume that anyone coming to this repo has any idea how to work with jsonnet, or even prometheus mixins. The design doc is a bit too high-level and verbose to tell a developer how to modify this structure.

We're also missing and explanation of what the various files in this directory are for. For instance, why are there both `services-external.libsonnet` and `services.libsonnet` when they appear to be identical?

Comment (anmolsachan) [reply metrics/mixin/README.md:30]: @jarrpa Its a new language and its not just us using it, It's a standard in Kubernetes monitoring world. Eg https://github.com/openshift/cluster-monitoring-operator/, https://github.com/kubernetes-monitoring/kubernetes-mixin. Just like I had to learn Golang to contribute to the OCS-operator, Jsonnet is a requirement to learn here. Also in OCS operator we have two active contributors for this, me and @umangachapagain and will be mostly. If more contributors have to contribute, they will have to learn.

Regarding the documentation, it's not as simple to get it, one has to do the whole tutorial to learn. Me and @umangachapagain have been through the pain. 

Regarding keeping two different files, I will give you an example: https://github.com/rook/rook/tree/master/cluster/examples/kubernetes/ceph/monitoring. If you have a look at the files in this repo you will see that external mode files have way lesser alerts than internal as they are no applicable there. 
We will have the same situation in OCS-operator as well, as the list of the alerts grows. Also, the OCS-operator alerts are not identical, there is a slight difference in the message and description fields which is because of the difference between the implementation of the internal and external mode of cephobjectstore.

Comment (jarrpa) [reply metrics/mixin/README.md:30]: This is worlds different than just knowing Golang. But even there, I have a todo item to improve the on-boarding documentation so people can get started with minimal working knowledge in Go. Best I can tell, Jsonnet doesn't have nearlyt he same market penetration as Go either, so you can't just assume people will be able to pick it up with the same ease. I don't want to skip that here, especially because *I* need to know what's going on before I can approve anything.

If you have the relevant Jsonnet tutorials, link those in the README. Any maintainer who is neither you or Umanga should be able to look at this PR and understand what is happening. Links to the projects you just referenced probably wouldn't hurt, either.

File: Makefile
Comment (umangachapagain) [inline Makefile:35]:
```suggestion
	gen-latest-alerts \
```

Just so everyone can understand

File: metrics/mixin/Makefile
Comment (umangachapagain) [inline metrics/mixin/Makefile:1]: Let's link it to the main Makefile so we can run these from root.

File: Makefile
Comment (umangachapagain) [inline Makefile:93]: add lint checks to ensure we generate valid rules.

File: metrics/mixin/README.md
Comment (umangachapagain) [inline metrics/mixin/README.md:5]:
```suggestion
The scope of this directory is to provide OCS specific Prometheus rule files using Monitoring Mixin.
```

File: metrics/mixin/alerts/services-external.libsonnet
Comment (umangachapagain) [inline metrics/mixin/alerts/services-external.libsonnet:5]:
```suggestion
        name: 'external-cluster-services-alert.rules',
```

File: metrics/mixin/alerts/services-external.libsonnet
Comment (umangachapagain) [inline metrics/mixin/alerts/services-external.libsonnet:18]: You can add endpoint details and state from the labels in metrics. That'll be more helpful.

File: metrics/mixin/alerts/services.libsonnet
Comment (umangachapagain) [inline metrics/mixin/alerts/services.libsonnet:20]: remove these. we don't use it

File: metrics/mixin/alerts/services.libsonnet
Comment (umangachapagain) [inline metrics/mixin/alerts/services.libsonnet:18]: add labels from metrics for more context.

File: metrics/mixin/alerts/services-external.libsonnet
Comment (umangachapagain) [inline metrics/mixin/alerts/services-external.libsonnet:20]: We can get rid of this.

File: metrics/mixin/build/example.jsonnet
Comment (umangachapagain) [inline metrics/mixin/build/example.jsonnet:1]: can we rename file to `internal.jsonnet`?

File: metrics/mixin/build/jsonnet/kube-prometheus.libsonnet
Comment (umangachapagain) [inline metrics/mixin/build/jsonnet/kube-prometheus.libsonnet:2]: Do we need this??

File: metrics/mixin/rules/rules.libsonnet
Comment (umangachapagain) [inline metrics/mixin/rules/rules.libsonnet:6]: Let's not keep it empty. 
Remove it and add when required.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

**Comment by umangachapagain** [comment]
What is `metrics/mixin/prometheus_rules.yaml` file? It shows no changes. Can we get rid of it?

File: Makefile
Comment (anmolsachan) [reply Makefile:35]: At the end of the days, its not just alerts, it will contain rules as well. By definition, they are together called Promethues Rules, that's why prometheus-rules-yamls, as rule yamls are the ones deployed.

File: metrics/mixin/Makefile
Comment (anmolsachan) [reply metrics/mixin/Makefile:1]: ack

File: Makefile
Comment (anmolsachan) [reply Makefile:93]: ack

File: metrics/mixin/README.md
Comment (anmolsachan) [reply metrics/mixin/README.md:5]: ack

File: metrics/mixin/alerts/services-external.libsonnet
Comment (anmolsachan) [reply metrics/mixin/alerts/services-external.libsonnet:5]: ack

File: metrics/mixin/alerts/services-external.libsonnet
Comment (anmolsachan) [reply metrics/mixin/alerts/services-external.libsonnet:18]: Makes sense

File: metrics/mixin/alerts/services.libsonnet
Comment (anmolsachan) [reply metrics/mixin/alerts/services.libsonnet:20]: We do, Its required for UI

File: metrics/mixin/alerts/services.libsonnet
Comment (anmolsachan) [reply metrics/mixin/alerts/services.libsonnet:18]: ack

File: metrics/mixin/alerts/services-external.libsonnet
Comment (anmolsachan) [reply metrics/mixin/alerts/services-external.libsonnet:20]: Yeah we can, but let's keep all the alerts related to OCS same. These will differ from the rook ones then, might create confusion.

File: metrics/mixin/build/example-external.jsonnet
Comment (anmolsachan) [reply metrics/mixin/build/example-external.jsonnet:1]: sure

File: metrics/mixin/build/example.jsonnet
Comment (anmolsachan) [reply metrics/mixin/build/example.jsonnet:1]: sure

File: metrics/mixin/build/jsonnet/kube-prometheus-external.libsonnet
Comment (anmolsachan) [reply metrics/mixin/build/jsonnet/kube-prometheus-external.libsonnet:2]: No, will remove

File: metrics/mixin/build/jsonnet/kube-prometheus.libsonnet
Comment (anmolsachan) [reply metrics/mixin/build/jsonnet/kube-prometheus.libsonnet:2]: No, will remove

File: metrics/mixin/rules/rules.libsonnet
Comment (anmolsachan) [reply metrics/mixin/rules/rules.libsonnet:6]: Cool

**Comment by anmolsachan** [comment]
> What is `metrics/mixin/prometheus_rules.yaml` file? It shows no changes. Can we get rid of it?

Done !.

**Comment by jarrpa** [comment]
With #739 merged, please update this one and we'll merge it straightaway.

**Comment by anmolsachan** [comment]
@jarrpa The PR is rebased. Thanks.

**Review by jarrpa — APPROVED** [review]

**Pull request merged**
